src/tests/unit/utils.ts

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+import { expect } from "chai";
+import { AppStoreState } from "../../app/stores";
+import { deepClone } from "../../core";
+import { DefaultAttributes } from "../../core/prototypes";
+
+export function makeDefaultAttributes(state: AppStoreState) {
+  const defaultAttributes: DefaultAttributes = {};
+  const { elements } = state.chart;
+  for (let i = 0; i < elements.length; i++) {
+    const el = elements[i];
+    defaultAttributes[el._id] = deepClone(
+      state.chartState.elements[i].attributes
+    );
+  }
+  return defaultAttributes;
+}
+
+/** Test if a deep equals b with tolerance on numeric values */
+export function expect_deep_approximately_equals(
+  a: any,
+  b: any,
+  tol: number,
+  context?: any
+) {
+  try {
+    if (a == null && b == null) {
+      return;
+    }
+    if (a == null || b == null) {
+      // If either of a, b is null/undefined
+      expect(a).equals(b, `${JSON.stringify(context, null, "")}`);
+    } else if (typeof a == "object" && typeof b == "object") {
+      if (a instanceof Array && b instanceof Array) {
+        // Both are arrays, recursively test for each item in the arrays
+        expect(a.length).to.equals(
+          b.length,
+          `${JSON.stringify(context, null, "")}`
+        );
+        for (let i = 0; i < a.length; i++) {
+          expect_deep_approximately_equals(a[i], b[i], tol, {
+            a,
+            b,
+          });
+        }
+      } else if (a instanceof Array || b instanceof Array) {
+        // One of them is an array, the other one isn't, error
+        throw new Error("type mismatch");
+      } else {
+        // Both are objects, recursively test for each key in the objects
+        const keysA = Object.keys(a).sort();
+        const keysB = Object.keys(b).sort();
+        expect(keysA).to.deep.equals(
+          keysB,
+          `${JSON.stringify(context, null, "")}`
+        );
+        for (const key of keysA) {
+          expect_deep_approximately_equals(a[key], b[key], tol, { a, b, key });
+        }
+      }
+    } else {
+      try {
```

Comment:
I'm not clear on what this try/catch block does. Please clarify with a comment.

Reply:
Numeric attributes in the `SVG` tree are stored as strings and we are trying to convert them to numbers and check for null. Such a check is needed to process the `SVG` path. Example `d="M 0,7.20843424 L 4.16179145,0 L 0,-7.20843424 L -4.16179145,0 Z"` It contains numbers and letters